megablocks/layers/mlp.py

```diff
@@ -173,47 +196,82 @@ def forward(ctx, x, w1, w2, topo):
         # pass in the backward pass to avoid materializing another
         # intermediate.
         ctx.shape = topo.shape
-        ctx.save_for_backward(
-            x, w1, w2, sdd_out.data,
-            topo.row_indices,
-            topo.column_indices,
-            topo.offsets,
-            topo.column_indices_t,
-            topo.offsets_t,
-            topo.block_offsets_t)
+        ctx.num_input_bits = num_input_bits
+        ctx.num_remat_bits = num_remat_bits
+        ctx.x_shape = x.shape
+        ctx.sdd_out_shape = sdd_out.data.shape
+        ctx.dtype = x.dtype
+        ctx.save_for_backward(*input_save_args, w1, w2, *topo_tensors)
         return dsd_out
 
     @staticmethod
     @torch.cuda.amp.custom_bwd
     def backward(ctx, ddsd_out):
-        x, w1, w2 = ctx.saved_tensors[:3]
-        sdd_out = stk.Matrix(ctx.shape, *ctx.saved_tensors[3:])
-
         if (not ctx.needs_input_grad[0] or
             not ctx.needs_input_grad[1] or
             not ctx.needs_input_grad[2]):
             raise ValueError("Expected all MLP inputs to need grad.")
 
+        # unpack saved tensors; ugly because quantizing changes tensor count
```

Comment:
Can you make this less ugly by putting the tensors for which there are potentially a variable number at the end of the saved_tensors? The particular part of the logic below that's hard to follow is the modification of saved_tensors.

Reply:
think this is prettier know but I'll leave it up to you as the reader whether this is "resolved"